rust/nix-util/src/context.rs

```diff
@@ -44,6 +83,49 @@ impl Drop for Context {
     }
 }
 
+#[macro_export]
+macro_rules! check_call {
+    ($f:path[$ctx:expr $(, $arg:expr)*]) => {
+        {
+            let ctx : &mut $crate::context::Context = $ctx;
+            let ret = $f(ctx.ptr() $(, $arg)*);
+            match ctx.check_err() {
+                Ok(_) => Ok(ret),
+                Err(e) => {
+                    ctx.clear();
+                    Err(e)
+                }
+            }
+        }
+    }
+}
+
+pub use check_call;
+
+// TODO: Generalize this macro to work with any error code or any error handling logic
```

Comment:
```suggestion
// TODO: Generalize this macro to work with any error code or any error handling logic. This is best done after https://github.com/NixOS/nix/issues/10635 is resolved.
```

I keep forgetting this is why we haven't done this already :sweat_smile: